File: examples/sguequierre/steps/doc-analyzer.step.py
# steps/doc-analyzer.step.py
import re
from typing import Dict, Any, List

# Configuration section
config = {
    "type": "event",
    "name": "doc-analyzer",
    "subscribes": ["documentation-analysis-started"],
    "emits": ["doc-gaps-identified"],
    "flows": ["documentation-guardian"]
}

# Handler function
def handler(payload, context):
    logger = context.logger
    emit = context.emit
    
    logger.info(f"Analyzing documentation for {payload['file']}")
    
    file_content = payload['content']
    file_path = payload['file']
    file_extension = file_path.split('.')[-1]
    
    # Different analysis strategies based on file type
    if file_extension in ['js', 'ts']:
        doc_gaps = analyze_js_ts(file_content)
    elif file_extension == 'py':
        doc_gaps = analyze_python(file_content)
    else:
        doc_gaps = analyze_generic(file_content)
    
    if not doc_gaps:
        logger.info(f"No documentation gaps found in {file_path}")
        return
    
    # Emit an event with the documentation gaps
    emit({
        "topic": "doc-gaps-identified",
        "data": {
            **payload,
            "gaps": doc_gaps
        }
    })

def analyze_js_ts(content):
    # Simple example - in reality, use an AST parser for better analysis
    functions = re.findall(r'(function\s+(\w+)|(\w+)\s*=\s*function|\(\s*\)\s*=>\s*{)', content)
    classes = re.findall(r'class\s+(\w+)', content)
    
    gaps = []
    
    # Check for missing JSDoc comments
    for match in functions:
        func_name = match[1] or match[2] or "anonymous"
        pattern = rf'/\*\*[\s\S]*?\*/\s*function\s+{func_name}|/\*\*[\s\S]*?\*/\s*{func_name}\s*='
        if not re.search(pattern, content):
            gaps.append({
                "type": "function",
                "name": func_name,
                "issue": "Missing JSDoc comment"
            })
    
    for class_name in classes:
        pattern = rf'/\*\*[\s\S]*?\*/\s*class\s+{class_name}'
        if not re.search(pattern, content):
            gaps.append({
                "type": "class",
                "name": class_name,
                "issue": "Missing class documentation"
            })
    
    return gaps

def analyze_python(content):
    # Check for missing docstrings
    functions = re.findall(r'def\s+(\w+)\s*\(', content)
    classes = re.findall(r'class\s+(\w+)', content)
    
    gaps = []
    
    for func_name in functions:
        pattern = rf'def\s+{func_name}\s*\([^)]*\):\s*([\'"]{{3}}|[\""]{{3}})'
        if not re.search(pattern, content):
            gaps.append({
                "type": "function",
                "name": func_name,
                "issue": "Missing docstring"
            })
    
    for class_name in classes:
        pattern = rf'class\s+{class_name}[^:]*:\s*([\'"]{{3}}|[\""]{{3}})'
        if not re.search(pattern, content):
            gaps.append({
                "type": "class",
                "name": class_name,
                "issue": "Missing class docstring"
            })
    
    return gaps

def analyze_generic(content):
    # Basic heuristic for other languages
    # In reality, you'd want language-specific analyzers
    lines = content.split('\n')
    gaps = []
    
    for i, line in enumerate(lines):
        if re.search(r'(function|def|class|interface|struct)\s+\w+', line) and i > 0:
            # Check if previous lines have comments
            has_comment = False
            for j in range(max(0, i-3), i):
                if '//' in lines[j] or '/*' in lines[j] or '#' in lines[j]:
                    has_comment = True
                    break
            
            if not has_comment:
                match = re.search(r'(function|def|class|interface|struct)\s+(\w+)', line)
                if match:
                    gaps.append({
                        "type": match.group(1),
                        "name": match.group(2),
                        "issue": "Missing documentation comment"
                    })
    
    return gaps

File: examples/sguequierre/steps/doc-analyzer.step.ts
// steps/doc-analyzer.step.ts
// Configuration section
export const config = {
  type: "event",
  name: "doc-analyzer",
  subscribes: ["file-content-fetched"],
  emits: ["doc-gaps-identified"],
  flows: ["documentation-guardian"]
};

// Type definitions
type Gap = {
  type: string;
  name: string;
  issue: string;
};

type Payload = {
  file: string;
  content: string;
  gaps?: Gap[];
};

type Context = {
  logger: {
    info: (message: string) => void;
    error: (message: string) => void;
  };
  emit: (event: { topic: string; data: any }) => Promise<void>;
};

// Handler function
export const handler = async (payload: Payload, context: Context): Promise<void> => {
  const { logger, emit } = context;
  
  logger.info(`Analyzing documentation for ${payload.file}`);
  
  const fileContent = payload.content;
  const filePath = payload.file;
  const fileExtension = filePath.split('.').pop() || '';
  
  // Different analysis strategies based on file type
  let docGaps: Gap[] = [];
  
  if (['js', 'ts'].includes(fileExtension)) {
    docGaps = analyzeJsTs(fileContent);
  } else if (fileExtension === 'py') {
    docGaps = analyzePython(fileContent);
  } else {
    docGaps = analyzeGeneric(fileContent);
  }
  
  if (!docGaps.length) {
    logger.info(`No documentation gaps found in ${filePath}`);
    return;
  }
  
  // Emit an event with the documentation gaps
  await emit({
    topic: "doc-gaps-identified",
    data: {
      ...payload,
      gaps: docGaps
    }
  });
};

function analyzeJsTs(content: string): Gap[] {
  // Can be improved: could use an AST parser for better analysis
  const functionRegex = /(function\s+(\w+)|(\w+)\s*=\s*function|\(\s*\)\s*=>\s*{)/g;
  const classRegex = /class\s+(\w+)/g;
  
  const gaps: Gap[] = [];
  
  // Check for missing JSDoc comments for functions
  let functionMatch;
  while ((functionMatch = functionRegex.exec(content)) !== null) {
    const funcName = functionMatch[2] || functionMatch[3] || "anonymous";
    const pattern = new RegExp(`/\\*\\*[\\s\\S]*?\\*/\\s*function\\s+${funcName}|/\\*\\*[\\s\\S]*?\\*/\\s*${funcName}\\s*=`);
    
    if (!pattern.test(content)) {
      gaps.push({
        type: "function",
        name: funcName,
        issue: "Missing JSDoc comment"
      });
    }
  }
  
  // Check for missing class documentation
  let classMatch;
  while ((classMatch = classRegex.exec(content)) !== null) {
    const className = classMatch[1];
    const pattern = new RegExp(`/\\*\\*[\\s\\S]*?\\*/\\s*class\\s+${className}`);
    
    if (!pattern.test(content)) {
      gaps.push({
        type: "class",
        name: className,
        issue: "Missing class documentation"
      });
    }
  }
  
  return gaps;
}

function analyzePython(content: string): Gap[] {
  // Check for missing docstrings
  const functionRegex = /def\s+(\w+)\s*\(/g;
  const classRegex = /class\s+(\w+)/g;
  
  const gaps: Gap[] = [];
  
  // Check functions for docstrings
  let functionMatch;
  while ((functionMatch = functionRegex.exec(content)) !== null) {
    const funcName = functionMatch[1];
    const pattern = new RegExp(`def\\s+${funcName}\\s*\\([^)]*\\):\\s*(['\\"]{3}|[\\"']{3})`);
    
    if (!pattern.test(content)) {
      gaps.push({
        type: "function",
        name: funcName,
        issue: "Missing docstring"
      });
    }
  }
  
  // Check classes for docstrings
  let classMatch;
  while ((classMatch = classRegex.exec(content)) !== null) {
    const className = classMatch[1];
    const pattern = new RegExp(`class\\s+${className}[^:]*:\\s*(['\\"]{3}|[\\"']{3})`);
    
    if (!pattern.test(content)) {
      gaps.push({
        type: "class",
        name: className,
        issue: "Missing class docstring"
      });
    }
  }
  
  return gaps;
}

function analyzeGeneric(content: string): Gap[] {
  // Basic heuristic for other languages
  const lines = content.split('\n');
  const gaps: Gap[] = [];
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const match = line.match(/(function|def|class|interface|struct)\s+(\w+)/);
    
    if (match && i > 0) {
      // Check if previous lines have comments
      let hasComment = false;
      
      for (let j = Math.max(0, i-3); j < i; j++) {
        if (lines[j].includes('//') || lines[j].includes('/*') || lines[j].includes('#')) {
          hasComment = true;
          break;
        }
      }
      
      if (!hasComment && match[1] && match[2]) {
        gaps.push({
          type: match[1],
          name: match[2],
          issue: "Missing documentation comment"
        });
      }
    }
  }
  
  return gaps;
}